src/GafferCycles/IECoreCyclesPreview/Renderer.cpp

```diff
@@ -1108,6 +1150,64 @@ class CyclesAttributes : public IECoreScenePreview::Renderer::AttributesInterfac
 
 			object->set_lightgroup( ccl::ustring( m_lightGroup.c_str() ) );
 
+			// Custom attributes.
+			ccl::vector<ccl::ParamValue> &attributes = object->attributes;
+			if( previousAttributes )
+			{
+				// Replace existing ones or remove ones that don't exist in the new attributes.
+				for( int i = attributes.size() - 1; i >= 0; i-- )
+				{
+					auto it = m_custom.find( attributes[i].name() );
+					if( it == m_custom.end() )
+					{
+						attributes.erase( attributes.begin() + i );
+					}
+					else
+					{
+						attributes[i] = SocketAlgo::setParamValue( it->first, it->second.get() );
+						if( !attributes[i].data() )
+						{
+							attributes.erase( attributes.begin() + i );
+							msg( Msg::Warning, "IECoreCycles::Renderer", boost::format( "Attribute \"%s\" applied to object \"%s\" has unsupported type \"%s\"." ) % it->first % object->name.string() % it->second->typeName() );
+						}
+					}
+				}
+
+				// Append any new attributes and skip replaced ones.
+				for( const auto &attr : m_custom )
+				{
+					auto it = previousAttributes->m_custom.find( attr.first );
+					if( it == previousAttributes->m_custom.end() )
```

Comment:
I think this will fail to apply an attribute that had an unsupported type before, but now has a supported type. Because it was unsupported before, it won't be in `attributes` already, so won't be updated in the "Replace existing ones" loop. But it will be in `previousAttributes->m_custom`, so this check will fail and we won't add it here either.

What's the upside to trying to make the minimal set of edits to `attributes`? Since it's just a plain `vector` with no signalling of the changes to Cycles, I can't see how Cycles would benefit from minimising edits. And since `ParamValue` has local storage for small types, I don't see any benefit in minimising assignments. On the other hand, the additional complexity makes the code more likely to contain bugs.

Is there any real downside to this simple proposal? :

- Store `m_custom` in a pre-converted `vector<ParamValue>`.
- In `applyObject()`, just do `object->attributes = m_custom`.
- If we really want to, avoid unnecessary assignments to `object->attributes` by storing a hash of `m_custom` and comparing against the hash for the previous attributes.

Reply:
Sounds good, yeah my original was along those lines sans the hashing comparison, that sounds much easier for sure.